README.md

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/zFactor)](https://cran.r-project.org/package=zFactor)
[![Travis-CI Build
Status](https://travis-ci.org/f0nzie/zFactor.svg?branch=master)](https://travis-ci.org/f0nzie/zFactor)
[![codecov](https://codecov.io/gh/f0nzie/zFactor/branch/master/graph/badge.svg)](https://codecov.io/gh/f0nzie/zFactor)

<!-- README.md is generated from README.Rmd. Please edit that file -->

# zFactor

Computational tools for chemical, petrochemical and petroleum engineers.
Calculates the error measures or deviation between different
correlations for gas compressibility compared to the original
Standing-Katz chart.

## Motivation

For the development of vertical lift performance (VLP) curves and other
petroleum engineering calculations, it is necessary to calculate
properties of hydrocarbons in mixtures, liquid and gaseous phases.
**Compressibility** is one of these hydrocarbon properties. It is
important to count with the apropriate correlation and know its range of
applicability in order to calculate other properties dependant of `z`.

There are several compressibility correlations. In this package, few of
them, or the most used in the industry, are being evaluated. There have
been extensive studies on compressibility correlations. This evaluation
is different in the sense that provides a graphical view of the range of
applicability as well statistical measures of the errors, besides that
the work has been written fully in `R` providing the analysis and
results for peer review and reproducibility purposes. All the code,
functions, notes, datasets are available via GitHub.

This analysis focus on sweet hydrocarbon gases. Compressibility
correlations for sour gases are little bit more complicated since they
show significant deviations from the curves in the Standing-Katz charts.
Effects of CO2, H2S and N2 have to be accounted for. In a future release
of `zFactor` correlations for *sour hydrocarbons gases* will possibly be
covered.

## Correlations used in this study

The correlations that are implemented in R for the package **zFactor**
are:

  - Beggs and Brill (BB) `[@Azizi2010]`

  - Hall and Yarborough (HY) `[@Hall1973]`

  - Dranchuk and Abou-Kassem (DAK) `[@Dranchuk1975]`

  - Dranchuk, Purvis and Robinson (DPR) `[@Dranchuk1973]`

  - A correlation by Shell Oil Company (SH) `[@Kumar2004]`,
    `[@bahadori]`, `[@Almeida2014]`, `[@Al-Anazi2011]`, `[@Azizi2010]`,
    `[@Mohamadi-Baghmolaei2015]`. The Shell correlation was found cited
    in two books and several papers (including the equation and
    constants), but the original paper or authors could not be
    identified.

  - A correlation developed with Artificial Neural Networks (Ann10) by
    Kamyab et al. `[@Kamyab2010]`

  - An explicit correlation by *I. Papp* `[@Papp1979]` mentioned in the
    comparative analysis paper by Gabor Takacs `[@Takacs1989]`. Papp’s
    paper is not available in English.

## Versions

  - In CRAN: `v 0.1.7`
  - In GitHub `development versions`

## Installation

I recommend installing from GitHub using `devtools`, that way you get
the latest and greatest version. CRAN release cycles allows updates
every one to two months.

You can install the latest version of `zFactor` from GitHub with:

``` r
# install.packages("devtools")
devtools::install_github("f0nzie/zFactor")
```

Or, if your prefer, from `CRAN`:

``` r
install.packages("zFactor")
```

## Usage

``` r
library(zFactor)

# get `z` values from the Standing-Katz chart
tpr <- c(1.2, 1.3, 1.5, 2.0, 3.0) 
ppr <- c(0.5, 1.5, 2.5, 3.5, 4.5, 5.5) 
getStandingKatzMatrix(ppr_vector = ppr, tpr_vector = tpr, 
                                 pprRange = "lp")
#>        0.5   1.5   2.5   3.5   4.5   5.5
#> 1.20 0.893 0.657 0.519 0.565 0.650 0.741
#> 1.30 0.916 0.756 0.638 0.633 0.684 0.759
#> 1.50 0.948 0.859 0.794 0.770 0.790 0.836
#> 2.00 0.982 0.956 0.941 0.937 0.945 0.969
#> 3.00 1.002 1.009 1.018 1.029 1.041 1.056

# calculate `z` using the Beggs-Brill correlation
z.BeggsBrill(pres.pr = 1.5, temp.pr = 2.0)
#> [1] 0.962902

# calculate `z` using the Hall-Yarborough correlation
z.HallYarborough(pres.pr = 1.5, temp.pr = 2.0)
#> [1] 0.9580002

# calculate `z` using the correlation Dranchuk-AbousKassem
ppr <- c(0.5, 1.5, 2.5, 3.5, 4.5, 5.5, 6.5) 
tpr <- c(1.05, 1.1, 1.7, 2) 
z.DranchukAbuKassem(ppr, tpr)
#>            0.5       1.5       2.5       3.5       4.5       5.5       6.5
#> 1.05 0.8300683 0.2837318 0.3868282 0.5063005 0.6239783 0.7392097 0.8521762
#> 1.1  0.8570452 0.4463987 0.4125200 0.5178068 0.6281858 0.7378206 0.8458725
#> 1.7  0.9681353 0.9128087 0.8753784 0.8619509 0.8721085 0.9003962 0.9409634
#> 2    0.9824731 0.9551087 0.9400752 0.9385273 0.9497137 0.9715388 1.0015560

# We do the same with the Dranchuk-Purvis-Robinson correlation
# but we change the values of the isotherms `Tpr`
tpr <- c(1.4, 1.4, 1.6, 1.7, 1.8) 
z.DranchukPurvisRobinson(pres.pr = ppr, temp.pr = tpr)
#>           0.5       1.5       2.5       3.5       4.5       5.5       6.5
#> 1.4 0.9374909 0.8155126 0.7275722 0.7083606 0.7404910 0.7977524 0.8666456
#> 1.4 0.9374909 0.8155126 0.7275722 0.7083606 0.7404910 0.7977524 0.8666456
#> 1.6 0.9602585 0.8891708 0.8399510 0.8230070 0.8364902 0.8709715 0.9183151
#> 1.7 0.9677844 0.9121791 0.8752677 0.8630002 0.8743271 0.9033216 0.9440582
#> 1.8 0.9736897 0.9298230 0.9022519 0.8944207 0.9053965 0.9311395 0.9672850

# now we use the relative undocumented Shell correlation
z.Shell(ppr, tpr)
#>           0.5       1.5       2.5       3.5       4.5       5.5       6.5
#> 1.4 0.9357838 0.8097802 0.7271788 0.7040931 0.7315325 0.7907497 0.8641572
#> 1.4 0.9357838 0.8097802 0.7271788 0.7040931 0.7315325 0.7907497 0.8641572
#> 1.6 0.9613216 0.8863715 0.8331061 0.8120822 0.8224613 0.8570695 0.9073678
#> 1.7 0.9711067 0.9150837 0.8740757 0.8563697 0.8629757 0.8901157 0.9321262
#> 1.8 0.9794808 0.9395295 0.9097288 0.8964473 0.9015059 0.9233190 0.9584750

# the newly added Papp correlation
z.Papp(ppr, tpr)
#>           0.5       1.5       2.5       3.5       4.5       5.5       6.5
#> 1.4 0.9408336 0.8010648 0.7076525 0.6906870 0.7299354 0.7958370 0.8705316
#> 1.4 0.9408336 0.8010648 0.7076525 0.6906870 0.7299354 0.7958370 0.8705316
#> 1.6 0.9657774 0.8864710 0.8252564 0.8026787 0.8182983 0.8604984 0.9168234
#> 1.7 0.9736623 0.9138939 0.8665932 0.8471834 0.8578133 0.8919610 0.9406180
#> 1.8 0.9796287 0.9345928 0.8988524 0.8837788 0.8922563 0.9208289 0.9632719

# and finally the correlation Kamyab et al that uses Artificial Neural Networks
z.Ann10(ppr, tpr)
#>           0.5       1.5       2.5       3.5       4.5       5.5       6.5
#> 1.4 0.9367118 0.8179531 0.7301083 0.7058966 0.7360320 0.7938440 0.8650626
#> 1.4 0.9367118 0.8179531 0.7301083 0.7058966 0.7360320 0.7938440 0.8650626
#> 1.6 0.9607316 0.8909372 0.8413772 0.8186001 0.8303206 0.8669610 0.9174184
#> 1.7 0.9682749 0.9146453 0.8767457 0.8581919 0.8672123 0.8978116 0.9413442
#> 1.8 0.9758251 0.9330673 0.9033038 0.8900081 0.8983954 0.9253309 0.9638663
```

## Range of Applicability

How to interpret the colors? We use the [Mean Absolute Percentage
Error](https://en.wikipedia.org/wiki/Mean_absolute_percentage_error) or
`MAPE` to visualize how close the correlation follow the experimental
values of the Standing-Katz chart.

  - `Blue`: the MAPE is zero or near zero
  - `Yellow`: the MAPE is around 10 percent.
  - `Red`: the MAPE has reached 25%.
  - `Grey`: the mean absolute percentage error is above or way above
    25%.

You can see for yourself which correlation is more stable at different
ranges of pseudo-reduce pressures and temperatures.

#### Beggs-Brill

``` r
library(zFactor)
zFactor:::z.plot.range("BB",  interval = "fine")
```

![](man/figures/README-unnamed-chunk-5-1.png)<!-- -->

#### Hall-Yarborough

``` r
zFactor:::z.plot.range("HY",  interval = "fine")
```

![](man/figures/README-unnamed-chunk-6-1.png)<!-- -->

``` r
zFactor:::z.plot.range("HY",  interval = "coarse")
```

![](man/figures/README-unnamed-chunk-6-2.png)<!-- -->

#### Dranchuk-AbouKassem

``` r
zFactor:::z.plot.range("DAK", interval = "fine")
```

![](man/figures/README-unnamed-chunk-7-1.png)<!-- -->

#### Dranchuk-Purvis-Robinson

``` r
zFactor:::z.plot.range("DPR", interval = "fine")
```

![](man/figures/README-unnamed-chunk-8-1.png)<!-- -->

#### Shell (Shell Oil Company)

``` r
zFactor:::z.plot.range("SH",  interval = "fine")
```

![](man/figures/README-unnamed-chunk-9-1.png)<!-- -->

#### Ann10 (Neural Network, 10 neurons)

``` r
zFactor:::z.plot.range("N10", interval = "fine")
```

![](man/figures/README-unnamed-chunk-10-1.png)<!-- -->

#### Papp (explicit)

``` r
zFactor:::z.plot.range("PP",  interval = "fine")
```

![](man/figures/README-unnamed-chunk-11-1.png)<!-- -->

## Comparative Analysis

The comparative analysis shows tables with different error measurements:

    MSE:   Mean Squared Error
    RMSE:  Root Mean Squared Error
    RSS:   Residual sum of Squares
    RMSLE: Root Mean Squared Logarithmic Error. Penalizes understimation.
    MAPE:  Mean Absolute Percentage Error = AARE
    MPE:   Mean Percentage error = ARE
    MAE:   Mean Absolute Error
    MAAPE: Mean Arc-tangent Absolute Percentage Error

> MAAPE was proposed by `[@Kim2016]`. Other authors consulted for the
> error measures are `[@Makridakis1993]`, `[@Chen2017]`,
> `[@Kobayashi2000]`, `[@Myttenaere2016]`, `[@Syntetos2013]`, and
> `[@Armstrong1992]`.

## What you can do with `zFactor`

  - Find `z` with any of the correlations provided
  - Get values from the Standing-Katz chart at any of the isotherms
    limited by the pseudo-reduced pressures (`Ppr`) digitized
  - Find what isotherms or pseudo-reduced temperatures (`Tpr`) are
    available from the Standing and Katz chart
  - Find what pseudo-reduced pressure points are available
  - Plot any of the Standing-Katz isotherms to view `z` in graphical
    form
  - Plot a tile chart to show the range of applicability of the
    correlations. In essence, get an idea of the accuracy by the
    location of the desired point at Tpr and Ppr
  - Get a statistics table when comparing any of the correlations to the
    Standing-Katz chart
  - Get a matrix of `z` values calculated from any correlation where the
    rows are the pseudo-reduced temperatures and the columns the
    pseudo-reduced pressures
  - Plot one or multiple isotherms

### Plotting a specific isotherm of SK chart

``` r
library(zFactor)
getStandingKatzCurve(tpr = 1.3, toView = FALSE, toSave = FALSE)
```

![](man/figures/README-unnamed-chunk-12-1.png)<!-- -->

``` r
getStandingKatzCurve(tpr = 1.05, pprRange = "lp", toView = FALSE, toSave = FALSE)
```

![](man/figures/README-unnamed-chunk-12-2.png)<!-- -->

### Ploting all the `Tpr` curves of Standing-Katz chart using `lapply`, `getStandingKatzData` and `data.table::rbindlist`:

``` r
library(zFactor)
library(ggplot2)
library(data.table)

tpr_vec <- c(1.05, 1.1, 1.2, 1.3, 1.5, 1.6, 1.7, 1.9, 2.0, 2.4, 2.6, 2.8, 3.0)
all_tpr2 <- (lapply(tpr_vec, function(x) getStandingKatzData(tpr = x)))
names(all_tpr2) <- tpr_vec

all_tpr_df <- data.table::rbindlist(all_tpr2, idcol = TRUE)
colnames(all_tpr_df)[1] <- "Tpr"

ggplot(all_tpr_df, aes(x=Ppr, y=z, group=Tpr, color=Tpr)) + 
    geom_line() +
    geom_point()
```

![](man/figures/README-unnamed-chunk-13-1.png)<!-- -->

### Build a table of statistical errors between a correlation and SK chart

``` r
library(ggplot2)
library(tibble)
library(dplyr)

# get all Tpr curves at low pressure (lp)
tpr_all <- getStandingKatzTpr(pprRange = "lp")
ppr <- c(0.5, 1.5, 2.5, 3.5, 4.5, 5.5, 6.5) 
sk_corr_all <- createTidyFromMatrix(ppr, tpr_all, correlation = "DPR")
grouped <- group_by(sk_corr_all, Tpr, Ppr)

smry_tpr_ppr <- summarise(grouped, 
          RMSE  = sqrt(mean((z.chart-z.calc)^2)), 
          MPE   = sum((z.calc - z.chart) / z.chart) * 100 / n(),
          MAPE  = sum(abs((z.calc - z.chart) / z.chart)) * 100 / n(), 
          MSE   = sum((z.calc - z.chart)^2) / n(), 
          RSS   = sum((z.calc - z.chart)^2),
          MAE   = sum(abs(z.calc - z.chart)) / n(),
          RMLSE = sqrt(1/n()*sum((log(z.calc +1)-log(z.chart +1))^2))
          )
as.tibble(smry_tpr_ppr)
#> # A tibble: 112 x 9
#>    Tpr     Ppr     RMSE    MPE   MAPE        MSE        RSS     MAE   RMLSE
#>    <chr> <dbl>    <dbl>  <dbl>  <dbl>      <dbl>      <dbl>   <dbl>   <dbl>
#>  1 1.05    0.5 0.000976  0.118  0.118    9.52e-7    9.52e-7 9.76e-4 5.33e-4
#>  2 1.05    1.5 0.0320   12.6   12.6      1.02e-3    1.02e-3 3.20e-2 2.52e-2
#>  3 1.05    2.5 0.0445   13.0   13.0      1.98e-3    1.98e-3 4.45e-2 3.26e-2
#>  4 1.05    3.5 0.0361    7.67   7.67     1.31e-3    1.31e-3 3.61e-2 2.43e-2
#>  5 1.05    4.5 0.0271    4.54   4.54     7.36e-4    7.36e-4 2.71e-2 1.68e-2
#>  6 1.05    5.5 0.0137    1.89   1.89     1.88e-4    1.88e-4 1.37e-2 7.92e-3
#>  7 1.05    6.5 0.00809   0.957  0.957    6.55e-5    6.55e-5 8.09e-3 4.38e-3
#>  8 1.1     0.5 0.00264   0.309  0.309    6.95e-6    6.95e-6 2.64e-3 1.42e-3
#>  9 1.1     1.5 0.0169    3.98   3.98     2.87e-4    2.87e-4 1.69e-2 1.18e-2
#> 10 1.1     2.5 0.0191    4.85   4.85     3.63e-4    3.63e-4 1.91e-2 1.36e-2
#> # … with 102 more rows
```

## Summary of MAPE statistics

This function shows a statistical summary of the Mean Absolute
Percentage Error of all the correlations. We are taking the mean,
maximum error, minimum eroor, the median and the mode in the whole range
of each opf the correlations.

``` r
library(zFactor)
stats_of_z.stats()
#>                 BB           HY          DAK         DPR         SH
#> mean    13.8862628  0.818657798  0.769721132  0.81237862  4.1530402
#> max    554.1872335 22.481448009 12.777908779 12.96594664 30.3942954
#> min      0.0153618  0.002422012  0.000609287  0.02057809  0.0231023
#> median   1.0402029  0.220154744  0.245336341  0.26714131  1.1092930
#> Mode     0.4281138  0.418077023  0.128870880  0.11769033  0.4389189
#>                N10           PP
#> mean   0.163515716  2.098729377
#> max    1.153159851 85.482666496
#> min    0.007209524  0.000507908
#> median 0.123437923  0.670820910
#> Mode   0.419773370  0.559623816
```

## Vignettes

The vignettes contain examples on the use and analysis of the various
correlations.

  - StandingKatz\_chart.Rmd
  - Beggs-Brill.Rmd
  - Hall-Yarborough.Rmd
  - Dranchuk-AbouKassem.Rmd
  - Dranchuk-Purvis-Robinson.Rmd
  - shell.Rmd
  - ANN.Rmd
  - Papp.Rmd

## Tests

There are unit tests for the correlations under tests/testthat.

## How the Standing-Katz chart was constructed

The Standing-Katz (SK) chart data has been read directly from a scanned
figure of the original plot drawn by Standing and Katz in 1951. The
software used to digitize the data is `graphClick` for the operating
system `osX` by Apple. This software has been tested and qualified by
scolars working on a similar task of digitizing data. `[@Rakap2016]`

Each one of the SK chart pseudo-reduce temperature curves has been
digitized. There are some curves where more points were taken to
describe the curvature better. Other curves are almost linear and few
points were necessary to define the curve.

## References

The following books and papers were consulted during the development of
this package:

\[1\] B. D. Al-Anazi, G. Pazuki, M. Nikookar, et al. “The prediction of
the compressibility factor of sour and natural gas by an artificial
neural network system”. In: *Petroleum Science and Technology* 29.4
(2011), pp. 325-336. DOI: 10.1080/10916460903330080. \<URL:
<http://doi.org/10.1080/10916460903330080>\>.

\[2\] J. C. de Almeida, J. A. Velásquez, and R. Barbieri. “A Methodology
for Calculating the Natural Gas Compressibility Factor for a
Distribution Network”. In: *Petroleum Science and Technology* 32.21
(2014), pp. 2616-2624. DOI: 10.1080/10916466.2012.755194. eprint:
<http://dx.doi.org/10.1080/10916466.2012.755194>. \<URL:
<http://dx.doi.org/10.1080/10916466.2012.755194>\>.

\[3\] J. Armstrong and F. Collopy. “Error measures for generalizing
about forecasting methods: Empirical comparisons”. In: *International
Journal of Forecasting* 8.1 (Jun. 1992), pp. 69-80. DOI:
10.1016/0169-2070(92)90008-w. \<URL:
<https://doi.org/10.1016/0169-2070(92)90008-W>\>.

\[4\] I. M. Azizi N. Behbahani R. “An efficient correlation for
calculating compressibility factor of natural gases”. In: *Journal of
Natural Gas Chemistry* Volume 19.Issue 6, 2010, (2010), pp. 642-645.
DOI: 10.1016/S1003-9953(09)60081-5. \<URL:
<http://doi.org/10.1016/S1003-9953(09)60081-5>\>.

\[5\] A. Bahadori. *Fluid Phase Behavior for Conventional and
Unconventional Oil and Gas Reservoirs*. ISBN 978-0-12-803437-8. Houston,
Texas: Gulf Publishing, 2016. \<URL:
<https://books.google.com/books?id=BwXeDAAAQBAJ>\>.

\[6\] C. Chen, J. Twycross, and J. M. Garibaldi. “A new accuracy measure
based on bounded relative error for time series forecasting”. In: *PLOS
ONE* 12.3 (Mar. 2017). Ed. by Z. Gao, p. e0174202. DOI:
10.1371/journal.pone.0174202. \<URL:
<http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0174202>\>.

\[7\] P. M. Dranchuk and H. Abou-Kassem. “Calculation of Z Factors For
Natural Gases Using Equations of State”. In: *Journal of Canadian
Petroleum Technology* (Jul. 1975). DOI: 10.2118/75-03-03. \<URL:
<https://doi.org/10.2118/75-03-03>\>.

\[8\] P. M. Dranchuk, R. Purvis, D. Robinson, et al. “Computer
calculation of natural gas compressibility factors using the Standing
and Katz correlation”. In: *Annual Technical Meeting*. Ed. by unknown.
Petroleum Society of Canada. 1973. DOI: 10.2118/73-112. \<URL:
<http://doi.org/10.2118/73-112>\>.

\[9\] K. R. Hall and L. Yarborough. “A new equation of state for
Z-factor calculations”. In: *Oil and Gas journal* 71.7 (1973),
pp. 82-92.

\[10\] M. Kamyab, J. H. Sampaio, F. Qanbari, et al. “Using artificial
neural networks to estimate the z-factor for natural hydrocarbon gases”.
In: *Journal of Petroleum Science and Engineering* 73.3 (2010),
pp. 248-257. DOI: 10.1016/j.petrol.2010.07.006. \<URL:
<http://doi.org/10.1016/j.petrol.2010.07.006>\>.

\[11\] S. Kim and H. Kim. “A new metric of absolute percentage error for
intermittent demand forecasts”. In: *International Journal of
Forecasting* 32.3 (Jul. 2016), pp. 669-679. DOI:
10.1016/j.ijforecast.2015.12.003. \<URL:
<https://doi.org/10.1016/j.ijforecast.2015.12.003>\>.

\[12\] K. Kobayashi and M. U. Salam. “Comparing Simulated and Measured
Values Using Mean Squared Deviation and its Components”. In: *Agronomy
Journal* 92.2 (2000), p. 345. DOI: 10.1007/s100870050043. \<URL:
<https://www.researchgate.net/publication/243112181_Comparing_Simulated_and_Measured_Values_Using_Mean_Squared_Deviation_and_Its_Components>\>.

\[13\] N. Kumar. “Compressibility factors for natural and sour reservoir
gases by correlations and cubic equations of state”. MA Thesis. Texas
Tech University, 2004. \<URL:
<https://ttu-ir.tdl.org/ttu-ir/handle/2346/1370>\>.

\[14\] S. Makridakis. “Accuracy measures: theoretical and practical
concerns”. In: *International Journal of Forecasting* 9.4 ( Dec. 1993),
pp. 527-529. DOI: 10.1016/0169-2070(93)90079-3. \<URL:
<https://www.sciencedirect.com/science/article/pii/0169207093900793>\>.

\[15\] M. Mohamadi-Baghmolaei, R. Azin, S. Osfouri, et al. “Prediction
of gas compressibility factor using intelligent models”. In: *Natural
Gas Industry B* 2.4 (2015), pp. 283-294. DOI:
10.1016/j.ngib.2015.09.001. \<URL:
<http://doi.org/10.1016/j.ngib.2015.09.001>\>.


NEWS.md
## zFactor 0.1.9.9000
* Accepted by CRAN on 20190801.
* TODO: add DOI for supporting paper.

## zFactor 0.1.9
* 20190731
* Released to CRAN
* Tested with Travis and rhub. All tests passed.
* change all as.tibble to `as_tibble`
* fix bibliography for Papps (DOI unrecognized)
* remove .Rprofile from rsuite
* README now accessing bibliography from `doc/` folder
* Build vignettes creating a `doc/` folder with `devtools::build_vignettes()`
* Install package with vignettes creating a `inst/doc/` folder with `devtools::install(build_vignettes = TRUE)`

## v0.1.8.9000
* 20190529
* zFactor now part of rsuite project
* found problem building vignettes from zfactor-rsuite
* will release to master to make things easier

##  v0.1.8
* 20190325
* release to CRAN


## v0.1.7.9003
* 20190324
* moving out vignettes, problem with symbolic links (bibliography)
* add function n() to importFrom in stats.R
* 

## v0.1.7.9002
* 20171028
*

## v0.1.7.9001
* 20171024
* comment two lines in .travis.yml
* Try not to use sciencedirect long links to prevent errors in knitcitations. shorten the URLs with doi links. 
* add statistical papers for citation. file ./inst/doc/error_stats.bib
* add error reference to README

## v0.1.7.9000
* 20171023
* add title to VignetteIndexEntry in statistics.Rmd. Preventing link to work in GitHub pages.
* edit bibliography.bib with JabRef application. ./vignettes/bibliography.bib is a symbolic link of same file under ./inst/doc
* Takacs was added entry for local PDF file in bibliography.bib


## 20171022 0.1.7
* returned by CRAN with two notes
* fixed name by adding R. as a middle name. Same as previous release.
* make z.stats as donotrun to shrink time for examples
* See cran-comments.md

## 20171019 0.1.6.9012
* release to CRAN as 0.1.7
* add function MAAPE
* add boxplots to show outliers in correlation calculation
* new vignette statistics
* add unit tests up to coverage 89%
* improve couple of functions: convertStringToVector , 
* add function stats_of_z.stats 
* add three datasets to ./tests/testthat/data


## 20171018 0.1.6.9011
* add citations to Papp vignette
* add correlation titles to sub-sections in README
* update README with references to new authors
* add two authors to bibliography: Takacs and Papp

## 20171018 0.1.6.9010
* change wrong Papay name to Papp in README, _pkgdown.yaml, vignette, util.R
* refresh site sending develop ./docs to branch `gh-pages using `git subtree`  

## 20171018 0.1.6.9009
* use a different git workflow
* delete unused tags in remote and local
* use "git subtree push --prefix docs origin gh-pages" in the Linux VM to send ./docs to remote branch gh-pages

## 20171016, 0.1.6.9008
* add vignette for Papay
* add function z.Papay
* fix Beggs-Brill vignette
* remove blank lines separating the badges
* TODO: Fix "Beggs-Brill Correlation". object --> sk_corr_3
* TODO: add Papp's correlation


## 20171016, 0.1.6.9007
* add travis.yml
* add code coverage with covr
* add badges to README
* rename couple of Rda files in ./data

## 20171010, 0.1.6.9006
* fix Rmd for Standing-Katz notebook

## 20171008, 0.1.6.9005
* small changes in the write-up of the README
* fix case in multiplotStandingKatz() when a vector of length 1 is specified
* ignore extra files in .Rbuildignore
* add more details and improve titles of several function. First group in "Standing-Katz chart" in customized menu.
* describe default values in parameters.
* move multiplot function to StandingKatz.R
* add some stops to functions

## 20171008, 0.1.6.9004
* customize navbar
* change menu to cerulean. 
* customize order of vignettes. 
* categorize functions.

## 20171008, 0.1.6.9003
* remove rdocumentation badge, it was static. changed to CRAN badge.

## 20171008, 0.1.6.9002
* add URL to description
* remove middle name from author

## 20171008, 0.1.6.9001
* Rebuild site after updates
* minor cosmetic changes in README
* add a link to wikipedia (MAPE)

## 20171008, 0.1.6.9000
* correct misspelled "hyddrocarbons". Noted by CRAN.
* resubmit to CRAN
* Build the package with `pkgdown`
* change author in description to use new way of adding authors
* tested offline and website running
* reinstalled package `rlang` to fix error during `pkgdown` installation


## 20170703, 0.1.6
* to be released to CRAN.
* CRAN found a problen in an URL in Shell.html file.
* Fixed problem by changing to a faster link. Previous link was loading a PDF from a website.
* remove bibliography.bib from inst/doc and letting devtools::build_vignettes() doing it instead. Necessary to add file `vignettes/.install_extras` with one line calling `bibliography.bib`. That way the file gets copied to inst/doc

## 20170702, 0.1.5
* to be released to CRAN. Problem with HY.rda
* add citations to README
* add citations to correlation notebooks
* use knitcitations
* add examples for all public functions
* improve description of package clarifying it is for sweet gases only
* update bibliography.bib. remove abstracts. ordered alphabetically.

## 20170629, 0.1.4.9016
* add more citations from JabRef
* note on sweet gases for now.
* as github_document

## 20170629, 0.1.4.9003
* all stat plots now show the name of the correlation
* stat tile plots can be drawn with Ppr `coarse` or `fine` interval
* deprecate `getCurvesDigitized` in favor of `getStandingKatzTpr`
* add new function `getStandingKatzPpr` which returns a short or long vector of Ppr intervals
* add new functions `isValid_correlation`,` get_z_correlations`
* add a dataframe `z_correlations` to contain z correlations info
* make title and subtitles dynamic in `z.plot.range`
* remove function `RMLSE` from stats dataframe. it was causing NaNs

## 20170629, 0.1.4.9002
* Finally fixed README images. Chaanged fig.path to man/figures/README-

## 20170629, 0.1.4.9001
* fix README images

## 20170628, 0.1.4.9000
* fix problem when passing a small vector or matrix to correlations.
* add new script stats.R that will calculate stats dataframe and plots

R/utils.R
# A dataframe with information about any of the compressibility correlation available
z_correlations <- data.frame(
    short = c("BB", "HY", "DAK", "DPR", "SH", "N10", "PP"),
    long = c("Beggs-Brill", "Hall-Yarborough", "Dranchuk-AbuKassem",
             "Dranchuk-Purvis-Robinson", "Shell", "Ann10", "Papp"),
    function_name = c("z.BeggsBrill", "z.HallYarborough", "z.DranchukAbuKassem",
                      "z.DranchukPurvisRobinson", "z.Shell", "z.Ann10", "z.Papp"),
    stringsAsFactors = FALSE
)


#' Get correlation information
#'
#' Get information about the correlation specifying for short name, long name or
#' the name of the function.
#'
#' @param how short: abbreviations; long: description; function: the name of the
#' correlation function
#' @export
#' @rdname get_z_correlations
#' @examples
#' # get the short name for the correlation
#' get_z_correlations(how = "short")
#'
#' # get the description for the correlation
#' get_z_correlations(how = "long")
#'
#' # get the name of the function assgined to the correlation
#' get_z_correlations(how = "function")
get_z_correlations <- function(how = "short") {
    # get correlation information. short: abbreviations; long: description
    #     function: the name of the correlation function
    if (how == "short")
        return(z_correlations[["short"]])
    if (how == "long")
        return(z_correlations[["long"]])
    if (how == "function")
        return(z_correlations[["function_name"]])
    else stop("wrong keyword")
}



#' split a long string to create a vector for testing
#'
#' @param str a contnuous long string to split as a vector
#' @export
#' @examples
#' convertStringToVector("1.05 1.10 1.20")
#' # result: "c(1.05, 1.1, 1.2)"
#' # now, you can paste the vector in your test
convertStringToVector <- function(str) {
    vs <- unlist(strsplit(str, " "))
    vs <- vs[lapply(vs, nchar) > 0]
    # prevent warning message:  introducing NAs by coercion
    # https://stackoverflow.com/a/14985152/5270873
    vn <- suppressWarnings(as.numeric(vs))
    if(all(is.na(vn))) vt <- paste(trimws(vs), collapse = ", ")
    else vt <- paste(vn, collapse = ", ")
    paste0('c(', vt, ')')
}


matrixToDataframe <- function(mat) {
    # convert a Ppr/Tpr matrix do a dataframe
    mat <- cbind(as.double(rownames(mat)), mat)  # new column for Tpr
    rownames(mat) <- NULL           # reset row names
    df <- as.data.frame(mat)        # dataframe
    names(df)[1] <- "Tpr"
    df
}


matrixWithCorrelation <- function(ppr_vector, tpr_vector, corr.Function) {
    # create a matrix using a z-factor correlation function and sapply
    co <- sapply(ppr_vector, function(x)
        sapply(tpr_vector, function(y) corr.Function(pres.pr = x, temp.pr = y)))

    if (length(ppr_vector) > 1 || length(tpr_vector) > 1) {
        co <- matrix(co, nrow = length(tpr_vector), ncol = length(ppr_vector))
        rownames(co) <- tpr_vector
        colnames(co) <- ppr_vector
    }
    co
}


combineCorrWithSK <- function(sk_df, co_df) {
    # combine correlation tidy DF with Standing-Katz tidy DF
    sk_cols <- ncol(sk_df)
    co_cols <- ncol(co_df)
    sk_tidy <- tidyr::gather(sk_df, "ppr", "z.chart", 2:sk_cols)
    co_tidy <- tidyr::gather(co_df, "ppr", "z.calcs", 2:co_cols)

    sk_co_tidy <- cbind(sk_tidy, z.calc = co_tidy$z.calcs)
    sk_co_tidy$dif <- sk_co_tidy$z.chart  - sk_co_tidy$z.calc
    colnames(sk_co_tidy)[1:2] <- c("Tpr", "Ppr")
    sk_co_tidy$Tpr <- as.character(sk_co_tidy$Tpr)
    sk_co_tidy$Ppr <- as.numeric(sk_co_tidy$Ppr)
    sk_co_tidy
}


#' Create a tidy table from Ppr and Tpr vectors
#'
#' @param ppr_vector a pseudo-reduced pressure vector
#' @param tpr_vector a pseudo-reduced temperature vector
#' @param correlation a z-factor correlation
#' @export
#' @examples
#' ppr <- c(0.5, 1.5, 2.5, 3.5)
#' tpr <- c(1.05, 1.1, 1.2)
#' createTidyFromMatrix(ppr, tpr, correlation = "DAK")
#' createTidyFromMatrix(ppr, tpr, correlation = "BB")
createTidyFromMatrix <- function(ppr_vector, tpr_vector, correlation) {
    msg <- "You have to provide a z-factor correlation: "
    msg_missing <- paste(msg, paste(get_z_correlations(), collapse = " "))
    if (missing(correlation)) stop(msg_missing)

    if (!isValid_correlation(correlation)) stop("Not a valid correlation.")

    zFunction <- get(z_correlations[which(z_correlations["short"] == correlation),
                                    "function_name"])

    sk_matrix <- getStandingKatzMatrix(ppr_vector, tpr_vector,
                                       pprRange = "lp")
    co_matrix <- matrixWithCorrelation(ppr_vector, tpr_vector,
                                       corr.Function = zFunction)

    sk_df <- matrixToDataframe(sk_matrix)
    co_df <- matrixToDataframe(co_matrix)

    sk_co_tidy <- combineCorrWithSK(sk_df, co_df)
    sk_co_tidy
}



#' Check if supplied correlation (three letter) is valid
#'
#' @param correlation a z-factor correlation
#' @export
isValid_correlation <- function(correlation) {
    # check if supplied correlation is valid
    valid_choices <- z_correlations[["short"]]      # get vector of correlations
    ifelse(correlation %in% valid_choices, TRUE, FALSE)
}






